dbms/src/Operators/Operator.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Core/Block.h>
+
+#include <memory>
+
+namespace DB
+{
+enum class OperatorStatus
+{
+    /// finish status
+    FINISHED,
+    CANCELLED,
+    /// spilling status
+    SPILLING,
+    /// waiting status
+    WAITING,
+    /// running status
+    // means that TransformOp/SinkOp needs to input a block to do the calculation,
+    // also used as an external state of `PipelineExec`.
+    NEED_INPUT,
+    // means that SourceOp/TransformOp output a block for used by subsequent operators.
+    // also used as an external state of `PipelineExec`.
+    HAS_OUTPUT,
+    // means that TransformOp handle the input block and output it.
+    PASS_THROUGH,
+};
+
+// TODO support operator profile info like `BlockStreamProfileInfo`.
+
+class Operator
+{
+public:
+    virtual ~Operator() = default;
+    virtual OperatorStatus await() { throw Exception("Unsupport"); }
+    virtual OperatorStatus spill() { throw Exception("Unsupport"); }
+};
+
+class SourceOp : public Operator
+{
+public:
+    virtual OperatorStatus read(Block & block) = 0;
+
+    virtual Block readHeader() = 0;
+
+    OperatorStatus await() override { return OperatorStatus::HAS_OUTPUT; }
+};
+using SourceOpPtr = std::unique_ptr<SourceOp>;
+
+class TransformOp : public Operator
+{
+public:
+    virtual OperatorStatus tryOutput(Block &) { return OperatorStatus::NEED_INPUT; }
+    virtual OperatorStatus transform(Block & block) = 0;
+
+    virtual void transformHeader(Block & header) { transform(header); }
+
+    OperatorStatus await() override { return OperatorStatus::NEED_INPUT; }
+};
+using TransformOpPtr = std::unique_ptr<TransformOp>;
+using TransformOps = std::vector<TransformOpPtr>;
+
+class SinkOp : public Operator
+{
+public:
+    virtual OperatorStatus prepare() { return OperatorStatus::NEED_INPUT; }
```

Comment:
It would be nice if comments add here to describe possible return status values in each situations. Ditto for other interfaces that return OperatorStatus.

Reply:
Do the comments at 
- https://github.com/pingcap/tiflash/pull/6520/files#diff-eff9da86fd05cd3ebffddf28ca658334df6cf9ec43f8b56601f9e22e6d4f9cb3R24-R98
- https://github.com/pingcap/tiflash/pull/6520/files#diff-a2e7fede76eed550de15cba59b5db9d995114f1c72b3c8976ee84de99da1152bR87-R138
meet expectations?